# Superstore Dataset — Pandas Data Exploration & Cleaning

**Objective:** Learn Python basics and perform basic data exploration and cleaning using Pandas.

**Dataset:** `Sample - Superstore.csv`

**Steps covered:**
1. Load the CSV into a DataFrame
2. Explore the data (head/tail, shape, columns, dtypes)
3. Handle missing values
4. Perform basic operations (filter rows, select columns)
5. Remove duplicates
6. Create a derived column (`total_amount = price * quantity`)
7. Save the cleaned dataset as a new CSV file


## 1. Load the CSV dataset into a Pandas DataFrame

In [1]:
import pandas as pd
import numpy as np

# The file uses Latin-1 encoding (contains extended-ASCII characters)
df = pd.read_csv('Sample_-_Superstore.csv', encoding='latin1')
print("Dataset loaded successfully.")
print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")


Dataset loaded successfully.
Rows: 9994, Columns: 21


## 2. Explore the data
Look at the first/last rows, shape, column names, and data types to understand what we're working with.

In [2]:
# First 5 rows
df.head()


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [3]:
# Last 5 rows
df.tail()


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
9989,9990,CA-2014-110422,1/21/2014,1/23/2014,Second Class,TB-21400,Tom Boeckenhauer,Consumer,United States,Miami,...,33180,South,FUR-FU-10001889,Furniture,Furnishings,Ultra Door Pull Handle,25.248,3,0.2,4.1028
9990,9991,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,FUR-FU-10000747,Furniture,Furnishings,Tenex B1-RE Series Chair Mats for Low Pile Car...,91.960,2,0.0,15.6332
9991,9992,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,TEC-PH-10003645,Technology,Phones,Aastra 57i VoIP phone,258.576,2,0.2,19.3932
9992,9993,CA-2017-121258,2/26/2017,3/3/2017,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,...,92627,West,OFF-PA-10004041,Office Supplies,Paper,"It's Hot Message Books with Stickers, 2 3/4"" x 5""",29.600,4,0.0,13.3200
9993,9994,CA-2017-119914,5/4/2017,5/9/2017,Second Class,CC-12220,Chris Cortes,Consumer,United States,Westminster,...,92683,West,OFF-AP-10002684,Office Supplies,Appliances,"Acco 7-Outlet Masterpiece Power Center, Wihtou...",243.160,2,0.0,72.9480


In [4]:
# Shape of the dataset (rows, columns)
df.shape


(9994, 21)

In [5]:
# Column names
df.columns.tolist()


['Row ID',
 'Order ID',
 'Order Date',
 'Ship Date',
 'Ship Mode',
 'Customer ID',
 'Customer Name',
 'Segment',
 'Country',
 'City',
 'State',
 'Postal Code',
 'Region',
 'Product ID',
 'Category',
 'Sub-Category',
 'Product Name',
 'Sales',
 'Quantity',
 'Discount',
 'Profit']

In [6]:
# Data types of each column
df.dtypes


Row ID             int64
Order ID             str
Order Date           str
Ship Date            str
Ship Mode            str
Customer ID          str
Customer Name        str
Segment              str
Country              str
City                 str
State                str
Postal Code        int64
Region               str
Product ID           str
Category             str
Sub-Category         str
Product Name         str
Sales            float64
Quantity           int64
Discount         float64
Profit           float64
dtype: object

In [7]:
# General summary: non-null counts, dtypes, memory usage
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   str    
 2   Order Date     9994 non-null   str    
 3   Ship Date      9994 non-null   str    
 4   Ship Mode      9994 non-null   str    
 5   Customer ID    9994 non-null   str    
 6   Customer Name  9994 non-null   str    
 7   Segment        9994 non-null   str    
 8   Country        9994 non-null   str    
 9   City           9994 non-null   str    
 10  State          9994 non-null   str    
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   str    
 13  Product ID     9994 non-null   str    
 14  Category       9994 non-null   str    
 15  Sub-Category   9994 non-null   str    
 16  Product Name   9994 non-null   str    
 17  Sales          9994 non-null   float64
 18  Quantity       9994

In [8]:
# Summary statistics for numeric columns
df.describe()


,Row ID,Postal Code,Sales,Quantity,Discount,Profit
count,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,4997.500000,55190.379428,229.858001,3.789574,0.156203,28.656896
std,2885.163629,32063.693350,623.245101,2.225110,0.206452,234.260108
min,1.000000,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,2499.250000,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,4997.500000,56430.500000,54.490000,3.000000,0.200000,8.666500
75%,7495.750000,90008.000000,209.940000,5.000000,0.200000,29.364000
max,9994.000000,99301.000000,22638.480000,14.000000,0.800000,8399.976000


## 3. Handle missing values
First identify how many missing values exist per column, then decide how to handle them
(fill with a sensible default, or drop rows/columns as appropriate).

In [9]:
# Count of missing values per column
missing_counts = df.isnull().sum()
missing_counts[missing_counts > 0] if missing_counts.sum() > 0 else "No missing values found in the dataset."


'No missing values found in the dataset.'

In [10]:
# This dataset happens to be complete (no missing values), but here is the general
# pattern used to handle missing data:

# Option A: Drop rows where ALL values are missing
df = df.dropna(how='all')

# Option B: Fill missing numeric values with the column median (example, applied defensively)
numeric_cols = df.select_dtypes(include=[np.number]).columns
for c in numeric_cols:
    if df[c].isnull().any():
        df[c] = df[c].fillna(df[c].median())

# Option C: Fill missing text/categorical values with 'Unknown'
text_cols = df.select_dtypes(include=['object']).columns
for c in text_cols:
    if df[c].isnull().any():
        df[c] = df[c].fillna('Unknown')

print("Missing values after cleaning:", df.isnull().sum().sum())


Missing values after cleaning: 0


/tmp/ipykernel_569/2280580117.py:14: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  text_cols = df.select_dtypes(include=['object']).columns


## 4. Perform basic operations (filter rows, select columns)

In [11]:
# Select a subset of columns
subset = df[['Order ID', 'Customer Name', 'Category', 'Sales', 'Quantity', 'Profit']]
subset.head()


,Order ID,Customer Name,Category,Sales,Quantity,Profit
0,CA-2016-152156,Claire Gute,Furniture,261.9600,2,41.9136
1,CA-2016-152156,Claire Gute,Furniture,731.9400,3,219.5820
2,CA-2016-138688,Darrin Van Huff,Office Supplies,14.6200,2,6.8714
3,US-2015-108966,Sean O'Donnell,Furniture,957.5775,5,-383.0310
4,US-2015-108966,Sean O'Donnell,Office Supplies,22.3680,2,2.5164


In [12]:
# Filter rows: orders in the 'Furniture' category with Sales > 500
furniture_high_sales = df[(df['Category'] == 'Furniture') & (df['Sales'] > 500)]
print(f"Found {len(furniture_high_sales)} Furniture orders with Sales > 500")
furniture_high_sales.head()


Found 467 Furniture orders with Sales > 500


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
10,11,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,...,90032,West,FUR-TA-10001539,Furniture,Tables,Chromcraft Rectangular Conference Tables,1706.1840,9,0.20,85.3092
24,25,CA-2015-106320,9/25/2015,9/30/2015,Standard Class,EB-13870,Emily Burns,Consumer,United States,Orem,...,84057,West,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,1044.6300,3,0.00,240.2649
27,28,US-2015-150630,9/17/2015,9/21/2015,Standard Class,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,...,19140,East,FUR-BO-10004834,Furniture,Bookcases,"Riverside Palais Royal Lawyers Bookcase, Royal...",3083.4300,7,0.50,-1665.0522


In [13]:
# Filter rows: profitable orders shipped via 'Same Day'
same_day_profitable = df[(df['Ship Mode'] == 'Same Day') & (df['Profit'] > 0)]
print(f"Found {len(same_day_profitable)} profitable Same Day orders")
same_day_profitable.head()


Found 444 profitable Same Day orders


,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
366,367,CA-2016-155516,10/21/2016,10/21/2016,Same Day,MK-17905,Michael Kennedy,Corporate,United States,Manchester,...,6040,East,OFF-BI-10002412,Office Supplies,Binders,Wilson Jones “Snap” Scratch Pad Binder Tool fo...,23.20,4,0.0,10.4400
367,368,CA-2016-155516,10/21/2016,10/21/2016,Same Day,MK-17905,Michael Kennedy,Corporate,United States,Manchester,...,6040,East,OFF-SU-10001225,Office Supplies,Supplies,Staple remover,7.36,2,0.0,0.1472
368,369,CA-2016-155516,10/21/2016,10/21/2016,Same Day,MK-17905,Michael Kennedy,Corporate,United States,Manchester,...,6040,East,OFF-ST-10002406,Office Supplies,Storage,Pizazz Global Quick File,104.79,7,0.0,29.3412
369,370,CA-2016-155516,10/21/2016,10/21/2016,Same Day,MK-17905,Michael Kennedy,Corporate,United States,Manchester,...,6040,East,FUR-BO-10002545,Furniture,Bookcases,"Atlantic Metals Mobile 3-Shelf Bookcases, Cust...",1043.92,4,0.0,271.4192
664,665,CA-2016-143308,11/4/2016,11/4/2016,Same Day,RC-19825,Roy Collins,Consumer,United States,Louisville,...,40214,South,OFF-FA-10000621,Office Supplies,Fasteners,"OIC Colored Binder Clips, Assorted Sizes",10.74,3,0.0,5.2626


## 5. Remove duplicates

In [14]:
# Check for fully duplicated rows
dup_count = df.duplicated().sum()
print(f"Duplicate rows found: {dup_count}")

# Remove them (keep the first occurrence)
df = df.drop_duplicates(keep='first')
print(f"Rows after removing duplicates: {len(df)}")


Duplicate rows found: 0


Rows after removing duplicates: 9994


## 6. Create a derived column: `total_amount = price * quantity`

The raw dataset doesn't include a standalone "price per unit" column — it has `Sales`
(the total sale amount for that line) and `Quantity`. To follow the exercise exactly, we:

1. Derive a `price` column as the unit price (`Sales / Quantity`)
2. Compute `total_amount = price * quantity`

As a sanity check, `total_amount` should equal the original `Sales` column.

In [15]:
# Derive unit price
df['price'] = df['Sales'] / df['Quantity']

# Derived column: total_amount = price * quantity
df['total_amount'] = df['price'] * df['Quantity']

# Sanity check: total_amount should match Sales (within floating point rounding)
df['sales_matches_total'] = np.isclose(df['total_amount'], df['Sales'], atol=0.01)
print("Rows where total_amount matches Sales:", df['sales_matches_total'].sum(), "/", len(df))

df[['Sales', 'Quantity', 'price', 'total_amount']].head()


Rows where total_amount matches Sales: 9994 / 9994


,Sales,Quantity,price,total_amount
0,261.9600,2,130.9800,261.9600
1,731.9400,3,243.9800,731.9400
2,14.6200,2,7.3100,14.6200
3,957.5775,5,191.5155,957.5775
4,22.3680,2,11.1840,22.3680


In [16]:
# Drop the helper verification column before saving
df = df.drop(columns=['sales_matches_total'])


## 7. Save the cleaned dataset as a new CSV file

In [17]:
output_path = 'Sample_Superstore_cleaned.csv'
df.to_csv(output_path, index=False)
print(f"Cleaned dataset saved to: {output_path}")
print(f"Final shape: {df.shape}")


Cleaned dataset saved to: Sample_Superstore_cleaned.csv
Final shape: (9994, 23)


## Summary

- **Original dataset:** 9,994 rows x 21 columns
- **Missing values:** none found (dataset was already complete); a general fill/drop
  pattern was demonstrated for robustness
- **Duplicates:** none found and removed defensively using `drop_duplicates()`
- **Filtering examples:** high-value Furniture orders, profitable Same Day shipments
- **Derived column:** `total_amount = price * quantity`, verified to reconstruct the
  original `Sales` column
- **Output:** cleaned dataset saved as `Sample_Superstore_cleaned.csv` with two new
  columns (`price`, `total_amount`) added

This notebook demonstrates the core Pandas workflow: **load → explore → clean →
transform → save** — the foundation for most real-world data analysis tasks.